python/metatensor-learn/metatensor/learn/data/dataset.py

```diff
@@ -0,0 +1,126 @@
+        if not isinstance(lazy_load, bool):
+            raise TypeError(f"`lazy_load` must be a bool, not {type(lazy_load)}")
+
+        # Check the number of samples (i.e. the list/dict length) is consistent
+        # between data items
+
+        num_samples = len(list(data.values())[0])
+        if not all(len(data_item) == num_samples for data_item in data.values()):
+            raise ValueError("Number of samples inconsistent between data items")
+
+        # If `sample_idxs` is not provided, use a continuous numeric range
+        if sample_idxs is None:
+            if any([isinstance(item, dict) for item in data.values()]):
+                raise ValueError(
+                    "Sample indices must be provided if passing data items as dicts"
+                )
+            sample_idxs = list(range(num_samples))
+
+        self._data_names = list(data.keys())
+        self._sample_idxs = sample_idxs
+        self._lazy_load = lazy_load
+
+        # Iterate through data items and samples. Check, load, and store if
+        # appropriate.
+        self._data = {name: {} for name in self._data_names}
+        for name, item in data.items():
+            if not isinstance(name, str):
+                raise TypeError(f"Data item name must be a str, not {type(name)}")
+
+            # Internally store each data item as a dict indexed by the sample
+            # index
+            if isinstance(item, list):
+                # Enumerate sample indices incase they are not passed as a
+                # continuous range
+                data[name] = {
+                    sample_idx: item[i] for i, sample_idx in enumerate(sample_idxs)
+                }
+
+            # Check `sample_idxs` is a subset of the indices present in the keys
+            # of the data item.
+            if isinstance(item, dict):
+                if not all(
+                    [sample_idx in list(item.keys()) for sample_idx in sample_idxs]
+                ):
+                    raise ValueError(
+                        f"`sample_idxs` must be a subset of the keys in the data item {name}"
+                    )
+            else:
+                raise TypeError(f"Data item must be a list or dict, not {type(item)}")
+
+            # Now only internally store the data for samples in `sample_idxs`
+            for sample_idx in sample_idxs:
+                if not lazy_load and callable(item[sample_idx]):  # eagerly load data
+                    self._data[name][sample_idx] = item[sample_idx]()
+                else:
+                    # Either already loaded, or a callable to be lazily loaded
+                    # later - do nothing yet
+                    self._data[name][sample_idx] = item[sample_idx]
+
+    def __len__(self) -> int:
+        """
+        Returns the length (i.e. number of samples) of the dataset
+        """
+        return len(self._sample_idxs)
+
+    def __getitem__(self, sample_idx: int, names: Optional[List[str]] = None) -> List:
```

Comment:
Yeah so I'm wondering whether this should return a dict or a list. It could be helpful to return a dict so that you have access to the sample indices, which is useful when getting a batch from a dataloader for instance. Thoughts?

Reply:
Is this the `OnDiskDataset` or you want to create a single Dataset for all the use cases? Because as far as I remember, the idea for to create a single Dataloader, but two datasets: `InMemoryDataset` and `OnDiskDataset`.

Reply:
Ok so I've changed it to return the sample indices as part of the batch, but fixed the type to be a List that is easily zip-able by the collate_fn. I've found this particularly useful when joining TensorMaps, as you can easily know the batch size (i.e. when the dataset isn't whole multiple of `batch_size` and there are remainders, for instance) and indices in that batch.

Reply:
Yeah, it's ok, but my question was - is it a prototype of the OnDiskDataset or a Dataset in general? Because, let say, if we have an input as a List[System] or Dict[str, System], it already seems to go out of the scope of the data types you assigned with the input.

Reply:
Resolving as now outdated - using `kwargs` instead. But yes, by passing callables we are providing a more general Dataset where the user can lazily load from disk some or all of the data